routing_plugin.go

```diff
@@ -104,3 +120,433 @@ func ReleaseRoutingPlugin(ctx context.Context) {
+}
+
+type nodeFeeInfo struct {
+	node     route.Vertex
+	capacity btcutil.Amount
+	fee      int64
+}
+
+// Enforce that lowToHighRoutingPlugin implements the RoutingPlugin interface.
+var _ RoutingPlugin = (*lowToHighRoutingPlugin)(nil)
+
+// buildPrivateChannels creates the private channel map from the passed route
+// hints. The code is taken and adapted from LND. Original source:
+// lnd/routing/payment_session_source.go.
+func buildPrivateChannels(routeHints [][]zpay32.HopHint,
+	target route.Vertex) map[route.Vertex][]*lndclient.ChannelEdge {
+
+	edges := make(map[route.Vertex][]*lndclient.ChannelEdge)
+
+	// Traverse through all of the available hop hints and include them in
+	// our edges map.
+	for _, routeHint := range routeHints {
+		// If multiple hop hints are provided within a single route
+		// hint, we'll assume they must be chained together and sorted
+		// in forward order in order to reach the target successfully.
+		for i, hopHint := range routeHint {
+			// In order to determine the end node of this hint,
+			// we'll need to look at the next hint's start node. If
+			// we've reached the end of the hints list, we can
+			// assume we've reached the target.
+			var toNode route.Vertex
+			if i != len(routeHint)-1 {
+				toNode = route.NewVertex(routeHint[i+1].NodeID)
+			} else {
+				toNode = target
+			}
+
+			fromNode := route.NewVertex(hopHint.NodeID)
+			// Finally, create the channel edges from the hop hint
+			// and add them to list of edges.
+			edgeFrom := &lndclient.ChannelEdge{
+				Node1:     fromNode,
+				Node2:     toNode,
+				ChannelID: hopHint.ChannelID,
+				Node1Policy: &lndclient.RoutingPolicy{
+					TimeLockDelta: uint32(
+						hopHint.CLTVExpiryDelta,
+					),
+					FeeBaseMsat: int64(
+						hopHint.FeeBaseMSat,
+					),
+					FeeRateMilliMsat: int64(
+						hopHint.FeeProportionalMillionths,
+					),
+				},
+			}
+			edges[fromNode] = append(edges[fromNode], edgeFrom)
+
+			edgeTo := &lndclient.ChannelEdge{
+				Node1:     toNode,
+				Node2:     fromNode,
+				ChannelID: hopHint.ChannelID,
+				Node2Policy: &lndclient.RoutingPolicy{
+					TimeLockDelta: uint32(
+						hopHint.CLTVExpiryDelta,
+					),
+					FeeBaseMsat: int64(
+						hopHint.FeeBaseMSat,
+					),
+					FeeRateMilliMsat: int64(
+						hopHint.FeeProportionalMillionths,
+					),
+				},
+			}
+			edges[toNode] = append(edges[toNode], edgeTo)
+		}
+	}
+
+	return edges
+}
+
+func getNodeInfo(ctx context.Context, lnd lndclient.LightningClient,
+	nodeID route.Vertex,
+	privateEdges map[route.Vertex][]*lndclient.ChannelEdge) (
+	*lndclient.NodeInfo, error) {
+
+	nodeInfo, err := lnd.GetNodeInfo(ctx, nodeID, true)
+	if err != nil {
+		// It's still possible that we should know this node from
+		// the provided route hints even though it is not part of the
+		// public graph. If we don't know any private channels then
+		// just return the error.
+		if _, ok := privateEdges[nodeID]; !ok {
+			return nil, err
+		}
+
+		err = nil
```

Comment:
This is a bit of an odd way to handle this to me. Could we rather match on `codes.NotFound` + check private edges else return error regardless?

Reply:
We _really_ shouldn't change error codes returned by lnd, so I think we're safe in this case. Also a lot of other naut/operator stuff will bork if we remove the `NotFound` status from `GetNodeInfo` 😅